src/Tabs.js

```diff
@@ -0,0 +1,233 @@
+'use strict';
+
+import core from 'metal';
+import templates from './Tabs.soy';
+import Component from 'metal-component';
+import Soy from 'metal-soy';
+
+/**
+ * UI Component to navigate through tabbed data.
+ */
+class Tabs extends Component {
+	/**
+	 * Adds a tab to the tabs array at the specified index if given or
+	 * appends it at the end.
+	 * @param {Object} tab
+	 * @param {number} index
+	 */
+	addTab(tab, index) {
+		if (core.isNumber(index)) {
+			this.tabs.splice(index, 0, tab);
+		} else {
+			this.tabs.push(tab);
+		}
+
+		this.tabs = this.tabs;
+	}
+
+	/**
+	 * Adds a tab to the tabs array at the specified index if given or
+	 * appends it at the end.
+	 * @param {string} label
+	 * @param {boolean} disabled
+	 * @param {number} index
+	 */
+	addTabByName(label, disabled, index) {
+		if (core.isString(label) && core.isBoolean(disabled)) {
+			let tab = {
+				label: label,
+				disabled: disabled
+			};
+
+			this.addTab(tab, index);
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	created() {
+		this.lastState_ = {
+			tab: this.tab
+		};
+
+		this.on(Tabs.Events.CHANGE_REQUEST, this.defaultChangeRequestFn_, true);
+	}
+
+	/**
+	 * Default `changeRequest` function, sets new state of tabs.
+	 * @param {EventFacade} event
+	 * @protected
+	 */
+	defaultChangeRequestFn_(event) {
+		this.setState_(event.state);
+	}
+
+	/**
+	 * Fires `changeRequest` event.
+	 * @param {Object} state
+	 * @protected
+	 */
+	dispatchRequest_(state) {
+		this.emit(
+			Tabs.Events.CHANGE_REQUEST,
+			{
+				lastState: this.lastState_,
+				state: state,
+				totalTabs: this.tabs.length
+			}
+		);
+	}
+
+	/**
+	 * Removes the tab at the given index from the tabs array.
+	 * @return {number} Returns the index of the first tab which is not disabled.
+	 */
+	findFirstAvailableIndex_() {
+		let index = -1;
+
+		for (let i = 0; i < this.tabs.length; i++) {
+			if ((!this.tabs[i].disabled || !core.isDef(this.tabs[i].disabled)) && !this.disabled) {
```

Comment:
Since you'll always ignore all tabs when `this.disabled` is `false`, it'd be best to do this check before the loop, to avoid it entirely. Something like:

```js
if (!this.disabled) {
  for (let i = 0; i < this.tabs.length; i++) {
    if (!this.tabs[i].disabled) {
      return i;
    }
  }
}
return -1;
```